backend/src/shortcuts.rs

```diff
@@ -18,107 +18,113 @@ pub struct Path {
 }
 
 impl Shortcuts {
-    pub fn new(map: &MapModel, neighbourhood: &Neighbourhood) -> Self {
+    pub fn new(map: &MapModel, neighbourhood: &Neighbourhood, cells: &Vec<Cell>) -> Self {
         let router_input_after = neighbourhood.shortcuts_router_input_after(map);
         let router_input_before = neighbourhood.shortcuts_router_input_before(map);
         let router_after = Router::new(&router_input_after, 1.0);
+
         let mut paths = Vec::new();
         let mut count_per_road = HashMap::new();
-        let mut checked_routes: HashSet<(RoadID, RoadID)> = HashSet::new();
-
-        for start_i in &neighbourhood.border_intersections {
-            let start_intersection = map.get_i(*start_i);
-            for start_r in &start_intersection.roads {
-                // It's not a "shortcut" unless it starts outside the interior and cuts through the interior.
-                if !neighbourhood.main_roads.contains(start_r) {
-                    continue;
-                }
-                for end_i in &neighbourhood.border_intersections {
-                    if start_i == end_i {
+
+        // A shortcut is contained within a cell, because cells can't cross modal filters or main
+        // roads, and neither can shortcuts. Only looking for candidate shortcuts in each cell is
+        // much faster than using all pairs of the neighbourhood's border intersections, because
+        // all the pairs belonging to different cells will need to go along or cross a main road to
+        // achieve a route.
+        for cell in cells {
+            for start_i in &cell.border_intersections {
+                let start_intersection = map.get_i(*start_i);
+                for start_r in &start_intersection.roads {
+                    // It's not a "shortcut" unless it starts outside the interior and cuts through
+                    // the interior.
+                    if !neighbourhood.main_roads.contains(start_r) {
                         continue;
                     }
-                    let end_intersection = map.get_i(*end_i);
-                    'next_road: for end_r in &end_intersection.roads {
-                        // It's not a "shortcut" unless it ends outside the interior after cutting through.
-                        if !neighbourhood.main_roads.contains(end_r) {
-                            continue;
-                        }
-
-                        // It's possible to repeat the route request when the same road is
-                        // connected to multiple border intersections
-                        if checked_routes.contains(&(*start_r, *end_r)) {
```

Comment:
To make sure I'm understanding: This `checked_routes.contains` call was removed because it is redundant, now that we only considers trips within a single cell.

Is that right?

Reply:
I removed it because of the reasoning in https://github.com/a-b-street/ltn/issues/124#issuecomment-2859379071 actually -- the same request could be formed from multiple directions, and to let turn restrictions at the border have an effect of shortcuts, we don't want to do this deduping.